scalding-core/src/main/scala/com/twitter/scalding/typed/CascadingBackend.scala

```diff
@@ -0,0 +1,268 @@
+package com.twitter.scalding.typed
+
+import cascading.flow.FlowDef
+import cascading.pipe.{ Each, Pipe }
+import cascading.tuple.{ Fields, TupleEntry }
+import com.twitter.scalding.TupleConverter.{ singleConverter, tuple2Converter }
+import com.twitter.scalding.TupleSetter.{ singleSetter, tup2Setter }
+import com.twitter.scalding._
+import com.twitter.scalding.serialization.{ CascadingBinaryComparator, OrderedSerialization, Boxed }
+import scala.collection.mutable.{ Map => MMap }
+import java.util.WeakHashMap
+
+object CascadingBackend {
+  import TypedPipe._
+
+  /**
+   * we want to cache renderings of some TypedPipe to Pipe so cascading
+   * will see them as the same. Without this, it is very easy to have
+   * a lot of recomputation. Ideally we would plan an entire graph
+   * at once, and not need a static cache here, but currently we still
+   * plan one TypedPipe at a time.
+   */
+  private[this] val pipeCache = new WeakHashMap[TypedPipe[Any], Map[Mode, CascadingPipe[Any]]]()
```

Comment:
Cool! I thought they'd be mutated by the cascading calls. Maybe we could encapsulate the mutable values and expose only safe methods? I worry if someone mutates the values by mistake since they are accessible as `CascadingPipe` fields.